igvc_perception/src/pointcloud_filter/fast_segment_filter/fast_segment_filter.cpp

```diff
@@ -0,0 +1,303 @@
+{
+  for (auto &seg : segments_)
+  {
+    sort(seg.second.prototype_points_.begin(), seg.second.prototype_points_.end());
+    Line curr_line = Line(config_.error_t);
+    for (const Prototype &pt : seg.second.prototype_points_)
+    {
+      if (!curr_line.attemptFitPoint(pt))
+      {
+        curr_line.is_ground_ = evaluateIsGround(curr_line);
+        seg.second.lines_.emplace_back(curr_line);
+        curr_line = Line(config_.error_t);
+        curr_line.attemptFitPoint(pt);
+      }
+    }
+    curr_line.is_ground_ = evaluateIsGround(curr_line);
+    if (curr_line.model_points_.size() > 1)
+    {
+      seg.second.lines_.emplace_back(curr_line);
+    }
+  }
+}
+
+void FastSegmentFilter::classifyPoints(pcl::PointCloud<velodyne_pointcloud::PointXYZIR> &ground_points,
+                                       pcl::PointCloud<velodyne_pointcloud::PointXYZIR> &nonground_points)
+{
+  for (const auto &[segment_id, segment] : segments_)
+  {
+    for (const auto &point : segment.raw_points_)
+    {
+      Line mapped_line;
+      double min_dist = -1;
+      for (const auto &line : segment.lines_)
+      {
+        for (const auto &pt : line.model_points_)
+        {
+          double distance = getDistanceBetweenPoints(point, pt.point_);
+          if (min_dist == -1 || distance < min_dist)
+          {
+            min_dist = distance;
+            mapped_line = line;
+          }
+        }
+      }
+      if (mapped_line.is_ground_)  // Removed max dist = nonground since was more pain than good
+      {
+        ground_points.push_back(point);
+      }
+      else
+      {
+        nonground_points.push_back(point);
+      }
+    }
+  }
+}
+
+int FastSegmentFilter::getSegIdFromPoint(const velodyne_pointcloud::PointXYZIR point)
+{
+  double angle = atan2(point.y, point.x) + M_PI;
+  return angle * config_.num_segments / (2 * M_PI);
+}
+
+double FastSegmentFilter::getDistanceFromPoint(const velodyne_pointcloud::PointXYZIR point)
+{
+  return std::hypot(point.x, point.y);
+}
+
+double FastSegmentFilter::getDistanceBetweenPoints(const velodyne_pointcloud::PointXYZIR point1,
+                                                   const velodyne_pointcloud::PointXYZIR point2)
+{
+  return std::hypot(point1.x - point2.x, point1.y - point2.y, point1.z - point2.z);
+}
+
+bool FastSegmentFilter::evaluateIsGround(Line &l)
+{
+  bool ground = false;
+  const auto above_intercept = [&](const Prototype &pt) { return pt.point_.z > config_.dist_t; };
+  bool all_below = std::all_of(l.model_points_.begin(), l.model_points_.end(), above_intercept);
+  if (all_below)
+  {
+    ground = true;
+  }
+  double dx = l.end_point_.point_.x - l.start_point_.point_.x;
+  double dy = l.end_point_.point_.y - l.start_point_.point_.y;
+  double dz = l.end_point_.point_.z - l.start_point_.point_.z;
+  double len_xy = std::hypot(dx, dy);
+  double slope = std::abs(dz) / len_xy;
+  ground = slope < config_.slope_t && l.end_point_.point_.z < config_.intercept_z_t;
+  return ground;
```

Comment:
It seems like the top two references of ground are dead and the actual use def for ground is only from 229 to 230?

Reply:
Oof wait lol I don't know how that happened. Fixed so it just returns true in the if statement.